edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+
+```rust
+pub enum Value {
+    Nothing,
+    Uuid(Uuid),
+    Str(String),
+    Bytes(Vec<u8>),
+    Int16(i16),
+    Int32(i32),
+    Int64(i64),
+    Float32(f32),
+    Float64(f64),
+    BigInt(BigInt),
+    // ... and so on
+}
+```
+
+Most variants of the `Value` enum correspond to a Rust type from the standard library, while some are from the `edgedb-protocol` crate and will have to be constructed. For example, this query expecting a `bigint` will return an error as it receives a `20` (an `i32`) but EdgeDB is expecting a `bigint`:
+
+```rust
+let query = "select <bigint>$0";
+let argument = 20;
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(format!("{query_res:?}").contains("expected std::int32"));
+```
+
+Instead, first construct a `BigInt` from the `i32` and pass that in as an argument:
+
+```rust
+use edgedb_protocol::model::BigInt;
+
+let query = "select <bigint>$0";
+let bigint_arg = BigInt::from(20);
+let query_res: Result<Value, _> = client.query_required_single(query, &(bigint_arg,)).await;
+assert_eq!(
+    format!("{query_res:?}"),
+    "Ok(BigInt(BigInt { negative: false, weight: 0, digits: [20] }))"
+);
+```
+
+## Casting inside the EdgeDB compiler
+
+EdgeDB requires arguments to have a cast in the same way that Rust requires a type declaration in function signatures. As such, arguments in queries are used as type specification for the EdgeDB compiler, not to cast from queries from the Rust side. Take this query as an example:
+
+```rust
+let query = "select <int32>$0";
+```
+
+This simply means "select an argument that must be an `int32`", not "take the received argument and cast it into an `int32`".
+
+As such, this will return an error:
+
+```rust
+let query = "select <int32>$0";
+let argument = 9i16; // Rust client will expect an int16
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(query_res
+    .unwrap_err()
+    .to_string()
+    .contains("expected std::int16"));
+```
+
+## Using JSON
+
+EdgeDB can cast any type to JSON with `<json>` so just sticking that in front of a query is enough to return the same object as JSON. This result can be turned into a struct using `serde` and `serde_json`.
+
+```rust
+#[derive(Debug, Deserialize)]
+pub struct Account {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let query = "with account := 
+    (insert Account {
+        username := <str>$0
+    }),
+    select <json>account {
+      username,
+      id
+    };";
+
+// We know there will only be one result so use query_single_json;
+// otherwise it will return a map of json
+let json_res = client
+    .query_single_json(query, &("SomeUserName",))
+    .await?
+    .unwrap();
+
+let as_account: Account = serde_json::from_str(&json_res)?;
+```
+
+## Using JSON with the `edgedb(json)` attribute
```

Comment:
Probably not for tutorial. Or show a usecase where `edgedb(json)` is better than returning object normally. In this example there is literally no difference.

Reply:
This one feels like trying to find a perfect alteranite example for edgedb(json) would just make it seem more important an attribute than it is, so opted to just remove it.